liferay-echo/liferay-portal
# LPS-160143  Add simpler API for getURLViewInContext

Motivation
=======
[LPS-160143](https://issues.liferay.com/browse/LPS-160143)
GetURLViewInContext relies on `liferayPortletRequest` and `liferayPortletResponse`, which are not available in fragments. In the majority of cases `themeDisplay` is the only data extracted from these parameters, thus providing a `themeDisplay` based implementation would speed up template migration from widget pages to content pages.

Proposed Solution
========
As discussed in [PTR-3216](https://issues.liferay.com/browse/PTR-3216), adding a new method that instead of `liferayPortletRequest` and `liferayPortletResponse` relies on `themeDisplay` could be a solution to migrate templates with minimal changes on the user's side.
I've created the new method as a default method so that backporting it would not be a breaking change
I've added 
```
throw new UnsupportedOperationException(
			"This method needs to be implemented");
```
to the default implementation because I couldn't find a way to make the logic work without a liferayPortletResponse for Base, CommerceOrder, Microblogs, JournalFolder and CalendarBooking. Please bare this in mind and share any concerns about this or ideas that might improve it.

Steps to Verify
========
1. Create a new Web Content Structure with a text field
1. Create a Template for it
```
<#assign
    journalArticleLocalService = serviceLocator.findService("com.liferay.journal.service.JournalArticleLocalService")
    assetEntryLocalService = serviceLocator.findService("com.liferay.asset.kernel.service.AssetEntryLocalService")
/>

<#assign
    article = journalArticleLocalService.getArticle(getterUtil.getLong(articleGroupId), .vars['reserved-article-id'].data) />
<#assign
    asset = assetEntryLocalService.getEntry('com.liferay.journal.model.JournalArticle', article.resourcePrimKey) />

<#if renderRequest??>
    <#assign
        urlViewInContext = asset.getAssetRenderer().getURLViewInContext(renderRequest, renderResponse, currentURL) />
<#else>
    <#assign
        urlViewInContext = asset.getAssetRenderer().getURLViewInContext(themeDisplay, currentURL) />
</#if>
<pre>${urlViewInContext}</pre>
```
1. Create a Widget Page with a Default Asset Publisher
2. Create a Web Content based on the new Structure and set the Widget Page as Display Page
1. Create a Content Page with an Asset Publisher and Publish it

*Expected behaviour:* urlViewInContext is available in both edit and view mode of the page.

Alternative Solutions that were discarded
========
Alternate solution [LPS-156378](https://issues.liferay.com/browse/LPS-156378) was discarded on https://github.com/liferay-echo/liferay-portal/pull/8549.
